test/cli-tests/no_permission_warn_test.sh

```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+. ./test/helper.sh
+
+test_install_dir="$test_fixtures_dir/no_permission_warn_test"
+
+function setUp()
+{
+	mkdir -p "$test_install_dir"
+	chmod -w "$test_install_dir"
+}
+
+function test_no_install_when_write_permissions_lacking()
+{
+	local output
+	local status
+	local inter="$(ruby-install --install-dir "$test_install_dir" ruby 2>&1)Z5Z5Z$?"
```

Comment:
I think this is too complex. `local output="$(...)"; ret=$?` should capture both successful and failure exit statuses.

Reply:
with this method, `ret` catches the exit status of the assignment rather than the command itself. the redirecting to stdout is also necessary for catching the error message

Reply:
```shell
output="$(exit -1)"; ret=$?
echo $ret
# 255
output="$(exit 0)"; ret=$?
echo $ret
# 0
```

Both `output="$(...)" || ret=$?` and `output="$(...)"; ret=$?` should capture the exit status, because the exit status of the previous command is kept in `$?`, until the current command or any embedded sub-command is completely executed.

Reply:
yes, if we dropped `local` modifier and declared output as global

Reply:
Oh right, `local` has it's own implicit return value. In that case the solution is simple, split the `local` definition from the assignment:

```shell
test_exit_status() {
	local output
	local ret

	output="$(exit -1)"
	ret=$?

	if [[ $ret == 255 ]]; then
		echo "exit status = -1"
	fi
}

test_exit_status
# exit status = -1
```